### PR TITLE
Exclude featuregate.release.openshift/tech-preview=true manifests

### PR DESCRIPTION
**Pull request by deads2k**
rebase and push of https://github.com/openshift/cluster-version-operator/pull/546

It still needs an auto-shutdown and probably (haven't yet looked) a few tests.

Implements https://github.com/openshift/enhancements/pull/957

### CODE REVIEW

File: pkg/payload/payload.go
Comment (wking) [inline pkg/payload/payload.go:224]: nit: this is asking about manifest exclusion based on some config knobs. I'm agnostic about whether the manifest argument goes before or after those knobs, but it probably shouldn't sit in the middle of the knobs.

File: pkg/cvo/cvo.go
Comment (wking) [inline pkg/cvo/cvo.go:227]: We'll want to have an actual `Context` input argument to this function before we land this, not the `TODO`, although obviously whatever's easiest for testing while you're still sketching.

Comment (deads2k) [reply pkg/cvo/cvo.go]: > We'll want to have an actual `Context` input argument to this function before we land this, not the `TODO`, although obviously whatever's easiest for testing while you're still sketching.

this has since been corrected and moved.

**Comment by JoelSpeed** [comment]
Have reviewed, have nothing to add, LGTM. Won't add the label as there are some linter errors that may want to be addressed

**Comment by alexander-demichev** [comment]
thanks!

File: pkg/featurechangestopper/featurechangestopper_test.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper_test.go:62]: [linter][1]:

```
pkg/featurechangestopper/featurechangestopper_test.go:62:15  errcheck  Error return value of `indexer.Add` is not checked
```

So:

```golang
if err := indexer.Add(...
  ...
); err != nil {
  t.Fatal(err)
}
```

or some such?

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/694/pull-ci-openshift-cluster-version-operator-master-golangci-lint/1463219181049614336#1:build-log.txt%3A16

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper.go:29]: nit: should this be "...returns a new TechPreviewChangeStopper." or "...a new TechPreviewNoUpgrade change stopper." or some such?

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper.go:100]: nit: There's not much going on in `runWorker` besides the comment here.  `Run` itself is fairly short and suite, so I dunno if it's worth a separate shim function just to get this comment out of `Run`.  Can we drop `runWorker` and inline its contents in `Run`?

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper.go:111]: From [the `Get` docs][1]:

> If `shutdown = true`, the caller should end their goroutine.

By returning `false` here, you're asking `runWorker` to stop, but `UntilWithContext` will just launch a new round, and seems like that would get into an unblocked hotloop.  Should we use [`PollWithContext`][2] instead, and then plumb the `done` boolean back up to the poller?  Something like:

```golang
wait.UntilWithContext(runContext, 30 * time.Second, func(ctx context.Context) (bool, error) {
  // hot loop blah blah your `runWorker` comment...
  for {
    if done, err := c.processNextWorkItem(ctx, shutdownFn); done || err != nil {
      return done, err
    }
  }
})
```
[1]: https://pkg.go.dev/k8s.io/client-go/util/workqueue#Type.Get
[2]: https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#PollWithContext

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper.go:87]: why launch a goroutine here?  The existing gorountine is just off to wait on `runContext` and then log the shutdown.  Can't you:

```golang
  wait...whatever...

  klog.Infof("Shutting down stop-on-techpreview-change controller")
}

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper.go:130]: nit: Formatting an `error` with `Errorf`, might as well [use `%w` instead of `%v` for compat with `Unwrap`][1].

[1]: https://pkg.go.dev/fmt#Errorf

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper.go:67]: nit: drop this line?  We'll fall through to the `return nil` two lines below.

File: pkg/featurechangestopper/featurechangestopper_test.go
Comment (wking) [inline pkg/featurechangestopper/featurechangestopper_test.go:77]: Testing this internal helper is convenient, but it seems like it would be about as convenient and also cover a bunch of vendored library code we're relying on if you exercised the public methods with something like:

```golang
c := New(tt.startingTechPreviewState, blaBlahNewFeatureGateInformer(indexer))
ctx, cancel := context.WithCancel(context.Background())
reaped := false
go func() {
  c.Run(ctx, shutdownFn)
  reaped = true
}()
time.Sleep(time.Minute)
cancel()
while !reaped {
  time.Sleep(time.Second)
}
```

File: pkg/payload/payload.go
Comment (wking) [inline pkg/payload/payload.go:228]: nit: The annotation _key_ is generically about feature gates.  The only tech-preview-ness that comes in is the value.  So `featureGateAnnotation`?  Or drop the local variable and just inline the key literal in the `if` line?

Comment (wking) [reply pkg/payload/payload.go:228]: Also, [from the enhancement][1]:

> If the value is not "TechPreviewNoUpgrade", the manifest is never created.

So:

```golang
if featureGate, ok := annotations["release.openshift.io/feature-gate"]; ok && !(includeTechPreview && featureGate == configv1.TechPreviewNoUpgrade) {
  return true
}
```

or something?

[1]: https://github.com/openshift/enhancements/pull/957/files#diff-6f9f82c1e111102d550554343fc634b0908acf22a0aa1490c0801f7ddf85b213R61

File: pkg/payload/payload_test.go
Comment (wking) [inline pkg/payload/payload_test.go:155]: nit: no need to dash these by hand; we can just write phrases like `"profile selection works"` and `testing` will slug them for us.

File: pkg/payload/payload_test.go
Comment (wking) [inline pkg/payload/payload_test.go:165]: Add two test cases [for][1]:

> If the value is not "TechPreviewNoUpgrade", the manifest is never created.

where you set the annotation to `DoesNotExist` and show that it's excluded regardless if `includeTechPreview`?

[1]: https://github.com/openshift/enhancements/pull/957/files#diff-6f9f82c1e111102d550554343fc634b0908acf22a0aa1490c0801f7ddf85b213R61

Comment (wking) [reply pkg/payload/payload_test.go:165]: Ah, you have these for `Other` already, but your exclusion sense on those tests diverges from your enhancement plan.

File: pkg/start/start.go
Comment (wking) [inline pkg/start/start.go:437]: if `err != nil`, we should probably at least log it.  Or should it be fatal?  "Random network hiccup, crashing out the CVO, hopefully our replacement process will have better luck" sounds better to me than "random network hiccup, I guess I will ignore all those TechPreviewNoUpgrade manifests I had been keeping an eye on for as long as I survive as a process".

The `context.TODO` and lack of `error` return from `NewControllerContext` both make me feel that this is a slightly odd place to put the Kube-API call.  But the existing split between `Options.Run` and `Options.NewControllerContext` seems pretty arbitrary to me.  I'm about even between either plumbing `ctx` in and `error` out of `NewControllerContext`, or with pulling this Kube-API call up into `Options.Run`.  Thoughts?

File: pkg/start/start_integration_test.go
Comment (wking) [inline pkg/start/start_integration_test.go:566]: I realize that these tests are all "we are not exercising TechPreviewNoUpgrade", but it feels a bit strange to be calling `NewControllerContext` two lines up, with its Kube-API call to check the feature gates and store the value for the CVO and stopper, and then here assume it's `false` instead of...  I guess two options:

a. Reach into `controllers` and pull `includeTechPreview` off of something.  Although I guess both the CVO and the stopper are holding the value in internal properties that we can't access from this package.  So maybe this one's not an option after all.
b. Shift the Kube-API call out of `NewControllerContext` and up into `Run`, and have the value passed into `NewControllerContext`.  Then here you can explicitly pass `false` into both `NewControllerContext` and `NewSyncWorker`.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (deads2k) [reply pkg/featurechangestopper/featurechangestopper.go:100]: > nit: There's not much going on in runWorker besides the comment here. Run itself is fairly short and sweet, so I dunno if it's worth a separate shim function just to get this comment out of Run. Can we drop runWorker and inline its contents in Run?


This is the standard structure from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure  .  I suggest keeping it the same to match standard kube controllers.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (deads2k) [reply pkg/featurechangestopper/featurechangestopper.go:111]: the `Get` only returns `quit = true` after it has been shutdown. this happens *after* the context is closed, which is plumbed to `UntilWithContext` in `Run`, so it doesn't restart.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (deads2k) [reply pkg/featurechangestopper/featurechangestopper.go:87]: > why launch a goroutine here? The existing gorountine is just off to wait on runContext and then log the shutdown. Can't you:

so that they stack neatly when you have more than one.  this allows parallel structures across every controller.  See the source link above.

File: pkg/featurechangestopper/featurechangestopper_test.go
Comment (deads2k) [reply pkg/featurechangestopper/featurechangestopper_test.go:77]: Based on past experience with controllers and unit tests, I recommend against that.  the time based approach is slow to run, so people don't and then when you tighten it fails in weird ways.  It also requires more plumbing in the general case.

File: pkg/payload/payload_test.go
Comment (deads2k) [reply pkg/payload/payload_test.go:155]: > nit: no need to dash these by hand; we can just write phrases like "profile selection works" and testing will slug them for us.

I'll defer to my reviewer, but that prevents a straight copy/find/paste flow in the code base.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [reply pkg/featurechangestopper/featurechangestopper.go]: Floating upstream with kubernetes/community#6266

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [reply pkg/featurechangestopper/featurechangestopper.go]: > ...this happens _after_ the context is closed...

This is based on some non-local assumptions, and while I don't expect this particular controller to be all that churny, implicit, non-local assumptions are generally brittle.  I've floated kubernetes/community#6267 to tidy up some shutdown handling in the upstream template.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [reply pkg/featurechangestopper/featurechangestopper.go]: I've floated kubernetes/community/pull/6267 to improve shutdown/irrecoverable-error handling in the upstream template.  Can you review and consider picking up any portions that make sense to you?

Comment (wking) [reply pkg/featurechangestopper/featurechangestopper.go]: Deferring this change; I'll file a follow-up CVO PR to pick up the stuff I'm suggesting in kubernetes/community#6267.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [reply pkg/featurechangestopper/featurechangestopper.go]: Deferring this change; I'll file a follow-up CVO PR to pick up the stuff I'm suggesting in kubernetes/community#6267.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [reply pkg/featurechangestopper/featurechangestopper.go]: Deferring this change; I'll file a follow-up CVO PR to pick up the stuff I'm suggesting in kubernetes/community#6267.

File: pkg/featurechangestopper/featurechangestopper.go
Comment (wking) [reply pkg/featurechangestopper/featurechangestopper.go]: fixed in fda77fef; thanks :)

File: pkg/payload/payload.go
Comment (wking) [reply pkg/payload/payload.go]: fixed in fda77fe; thanks :)

File: pkg/payload/payload_test.go
Comment (wking) [reply pkg/payload/payload_test.go]: fixed in fda77fe; thanks :)

And fair enough on copy-paste-ness, but the existing CVO pattern is unslugged values.  I'm fine with a larger pivot in this direction if we want to shift conventions...

File: pkg/payload/payload_test.go
Comment (wking) [reply pkg/payload/payload_test.go:165]: fixed in fda77fe; thanks :)

File: pkg/start/start.go
Comment (wking) [reply pkg/start/start.go]: fixed in fda77fe, which makes the error fatal and shifts the API call up into `Run`; thanks :)

File: pkg/start/start_integration_test.go
Comment (wking) [reply pkg/start/start_integration_test.go]: fixed in fda77fe; thanks :)  Pre-existing brittle-ness from the `SetSyncWorkerForTesting` flow can slide.

File: pkg/featurechangestopper/featurechangestopper_test.go
Comment (wking) [reply pkg/featurechangestopper/featurechangestopper_test.go]: I'm fine punting, but may come back to this later on.

**Review by wking — APPROVED** [review]
/lgtm

Good stuff; thanks :)

File: pkg/start/start.go
Comment (wking) [inline pkg/start/start.go:448]: oops, this API call needs to move up into `Run`, now that you're passing `includeTechPreview` into `NewControllerContext`.  [CI][1]:

```
pkg/start/start.go:448:21: no new variables on left side of :=
```

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/694/pull-ci-openshift-cluster-version-operator-master-images/1466836699643383808#1:build-log.txt%3A67

**Comment by LalatenduMohanty** [comment]
This PR also needs  corresponding documentation in https://github.com/openshift/enhancements/tree/master/dev-guide/cluster-version-operator/dev

File: pkg/start/start.go
Comment (wking) [reply pkg/start/start.go]: fixed in c323882e.

/lgtm

**Comment by wking** [comment]
[sandbox error][1] is orthogonal:

/override ci/prow/e2e-agnostic-upgrade

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/694/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-upgrade/1466841686981218304

**Comment by wking** [comment]
Looks like #703 snuck in in parallel, and now Tide is retesting the world against the new target commit, but I want to close out this epic and move on to other things, so skip the slow stuff:

/override ci/prow/e2e-agnostic
/override ci/prow/e2e-agnostic-operator

**Pull request merged**
